src/chain/chain-generic-numerator.cc

```diff
@@ -98,246 +92,269 @@ GenericNumeratorComputation::GenericNumeratorComputation(
+  alpha_[seq].Set(-std::numeric_limits<BaseFloat>::infinity());
+  alpha_[seq](0, 0) = 0.0;
+  alpha_[seq](0, num_states) = 0.0;
 }
 
 
+void GenericNumeratorComputation::CopySpecificPdfsIndirect(
+                                    int sequence_id,
+                                    const std::vector<MatrixIndexT> &indices,
+                                    Matrix<BaseFloat> *out) {
+  int num_sequences = supervision_.num_sequences;
+  int frames_per_sequence = supervision_.frames_per_sequence;
+  int num_pdfs = nnet_output_x_.NumCols();
+
+  const BaseFloat *starting_ptr = nnet_output_x_.RowData(sequence_id);
+  int view_stride = num_sequences * nnet_output_x_.Stride();
+  const CuSubMatrix<BaseFloat> sequence_view(starting_ptr,
+                                             frames_per_sequence,
+                                             num_pdfs,
+                                             view_stride);
+
+  std::vector<MatrixIndexT> indices_expanded(num_pdfs, -1);
+  copy(indices.begin(), indices.end(), indices_expanded.begin());
+
+  CuArray<MatrixIndexT> indices_gpu(indices_expanded);
+  CuMatrix<BaseFloat> single_sequence(frames_per_sequence,
+                                      sequence_view.NumCols());
+
+  single_sequence.CopyCols(sequence_view, indices_gpu);
+  CuSubMatrix<BaseFloat> sequence_pdfs_gpu(single_sequence,
+                             0, single_sequence.NumRows(),
+                             0, indices.size());
+
+  out->Resize(sequence_pdfs_gpu.NumCols(),
+             sequence_pdfs_gpu.NumRows());
+  out->CopyFromMat(sequence_pdfs_gpu, kTrans);
+}
+
 // The alpha computation for some 0 < t <= num_time_steps_.
-void GenericNumeratorComputation::AlphaGeneralFrame(int32 t) {
+BaseFloat GenericNumeratorComputation::AlphaGeneralFrame(int32 seq) {
   // Define some variables to make things nicer
   const int32
-      num_sequences = supervision_.num_sequences,
-      num_frames = supervision_.frames_per_sequence,
-      num_pdfs = exp_nnet_output_transposed_.NumRows(),
-      num_states = max_num_hmm_states_;
-  KALDI_ASSERT(t > 0 && t <= num_frames);
-
-  SubMatrix<double> this_alpha(alpha_.RowData(t), num_states,
-                               num_sequences, num_sequences);
-  const SubMatrix<double> prev_alpha(alpha_.RowData(t - 1), num_states + 1,
-                                     num_sequences, num_sequences);
-  // 'probs' is the matrix of pseudo-likelihoods for frame t - 1.
-  SubMatrix<BaseFloat> probs(exp_nnet_output_transposed_, 0, num_pdfs,
-                             (t - 1) * num_sequences, num_sequences);
-
-  for (int32 seq = 0; seq < num_sequences; seq++) {
-    double inv_arbitrary_scale = prev_alpha(num_states, seq);
+    num_sequences = supervision_.num_sequences,
+    num_frames = supervision_.frames_per_sequence;
+
+  KALDI_ASSERT(seq >= 0 && seq < num_sequences);
+  CopySpecificPdfsIndirect(seq, index_to_pdf_[seq], &(probs_[seq]));
+
+  SubMatrix<BaseFloat> alpha(alpha_[seq],
```

Comment:
I guess something that can make the code even nicer, is to have a separate `alpha_sum` variable (maybe `std::vector<Vector<BaseFloat> > alpha_sum`) which stores the alpha-sums. I guess it will simplify some parts of the code. 
If you want to merge, I can do this in a separate PR?

Reply:
Actually, I'm not even sure if alpha sums help (with numeric stability) in log-space?